### PR TITLE
feature-NXP-22798-apache-http-client-rest-tests

### PR DESCRIPTION
**Pull request by ataillefer**
See https://qa.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-ataillefer/48/, failed only because of a random failure in the CMIS tests

### CODE REVIEW

File: nuxeo-features/rest-api/nuxeo-rest-api-server/src/main/java/org/nuxeo/ecm/restapi/server/jaxrs/BatchUploadObject.java [hunk 1]
```diff
@@ -139,96 +138,110 @@ protected Response uploadNoTransaction(@Context HttpServletRequest request,
...
+        }
+        return Long.parseLong(contentLength);
+    }
+
+    protected void addStream(String uploadType, String batchId, String fileIdx, InputStream is, String fileName,
+            String mimeType, long uploadedSize, int chunkCount, int uploadChunkIndex, long fileSize)
+            throws IOException {
+        String uploadedSizeDisplay = uploadedSize > -1 ? uploadedSize + "b" : "unknown size";
+        if (UPLOAD_TYPE_CHUNKED.equals(uploadType)) {
+            log.debug(String.format("Uploading chunk [index=%d / total=%d] (%s) for file %s", uploadChunkIndex,
```
Comment (kevinleturc) [inline nuxeo-features/rest-api/nuxeo-rest-api-server/src/main/java/org/nuxeo/ecm/restapi/server/jaxrs/BatchUploadObject.java]: I'm wondering here if this code is called a lot or not, construction of log message could hit down performance. Maybe we can add a `isDebugEnabled()`, wdyt ?

File: nuxeo-features/nuxeo-features-test/src/test/java/org/nuxeo/test/runner/feature/RetryPostTest.java [hunk 2]
```diff
@@ -91,6 +91,7 @@ public void testRetryPostEnabled() throws Exception {
         // POST retry is enabled, the call should succeed
         ClientResponse response = webResource.post(ClientResponse.class);
         assertEquals(200, response.getStatus());
+        response.close();
```
Comment (efge) [inline nuxeo-features/nuxeo-features-test/src/test/java/org/nuxeo/test/runner/feature/RetryPostTest.java]: A try/finally would be better here.

File: nuxeo-features/rest-api/nuxeo-rest-api-test/src/main/java/org/nuxeo/ecm/restapi/test/BaseTest.java [hunk 3]
```diff
@@ -71,62 +71,74 @@
...
-        return client.resource("http://localhost:18090/api/v1/");
+    /**
+     * Returns a {@link WebResource} to perform calls on the given resource with the given credentials.
+     * <p>
+     * Uses the Apache HTTP client, more reliable and much more configurable than the one from the JDK.
+     *
+     * @since 9.3
+     */
+    protected WebResource getServiceFor(String resource, String username, String password) {
+        client = JerseyClientHelper.clientBuilder().setCredentials(username, password).build();
```
Comment (kevinleturc) [inline nuxeo-features/rest-api/nuxeo-rest-api-test/src/main/java/org/nuxeo/ecm/restapi/test/BaseTest.java]: Should you destroy the previous client when a call is made on `getServiceFor` ?

File: nuxeo-features/nuxeo-features-test/src/main/java/org/nuxeo/jaxrs/test/JerseyClientHelper.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    public static ApacheHttpClientBuilder clientBuilder() {
+        return new ApacheHttpClientBuilder();
+    }
+
+    /**
+     * Makes the given {@link ClientResponse} {@link AutoCloseable} by wrapping it in a {@link CloseableClientResponse}.
+     */
+    public static CloseableClientResponse toAutoCloseable(ClientResponse response) {
+        return new CloseableClientResponse(response.getStatusInfo(), getResponseHeaders(response),
+                response.getEntityInputStream(), response.getClient().getMessageBodyWorkers());
```
Comment (efge) [inline nuxeo-features/nuxeo-features-test/src/main/java/org/nuxeo/jaxrs/test/JerseyClientHelper.java]: I would move this method to the `CloseableClientResponse` class and call it `of`.

File: nuxeo-features/nuxeo-features-test/src/main/java/org/nuxeo/jaxrs/test/JerseyClientHelper.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+        private int connectTimeout;
+
+        private int connectionRequestTimeout;
+
+        private String username;
+
+        private String password;
+
+        ApacheHttpClientBuilder() {
```
Comment (efge) [inline nuxeo-features/nuxeo-features-test/src/main/java/org/nuxeo/jaxrs/test/JerseyClientHelper.java]: `public` ?

File: nuxeo-features/nuxeo-features-test/src/main/java/org/nuxeo/jaxrs/test/JerseyClientHelper.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
```
Comment (ataillefer) [inline nuxeo-features/nuxeo-features-test/src/main/java/org/nuxeo/jaxrs/test/JerseyClientHelper.java]: Yes, changing it to `protected` and for all the members too

File: nuxeo-features/rest-api/nuxeo-rest-api-server/src/main/java/org/nuxeo/ecm/restapi/server/jaxrs/BatchUploadObject.java [hunk 7]
```diff
@@ -139,96 +138,110 @@ protected Response uploadNoTransaction(@Context HttpServletRequest request,
...
+            FormData formData = new FormData(request);
+            Blob blob = formData.getFirstBlob();
+            if (blob == null) {
+                throw new NuxeoException("Cannot upload in multipart with no blobs");
+            }
+            if (!UPLOAD_TYPE_CHUNKED.equals(uploadType)) {
+                fileName = blob.getFilename();
+            }
+            // Don't change the mime-type if it was forced via the X-File-Type header
+            if (mimeType == null || mimeType.isEmpty()) {
```
Comment (ataillefer) [inline nuxeo-features/rest-api/nuxeo-rest-api-server/src/main/java/org/nuxeo/ecm/restapi/server/jaxrs/BatchUploadObject.java]: :+1:

File: nuxeo-features/rest-api/nuxeo-rest-api-server/src/main/java/org/nuxeo/ecm/restapi/server/jaxrs/BatchUploadObject.java (hunk 1 above)
Comment (ataillefer) [inline nuxeo-features/rest-api/nuxeo-rest-api-server/src/main/java/org/nuxeo/ecm/restapi/server/jaxrs/BatchUploadObject.java]: Old code but yes

File: nuxeo-features/rest-api/nuxeo-rest-api-test/src/main/java/org/nuxeo/ecm/restapi/test/BaseTest.java (hunk 3 above)
Comment (ataillefer) [inline nuxeo-features/rest-api/nuxeo-rest-api-test/src/main/java/org/nuxeo/ecm/restapi/test/BaseTest.java]: Right

**Comment by efge** [comment]
👍

**Pull request merged**